lib/nx_signal.ex

```diff
@@ -297,4 +481,151 @@ defmodule NxSignal do
+    * `fs` - Sampling frequency in Hz
+
+  ## Options
+    * `:max_mel` - the pitch for the last MEL bin before log scaling. Defaults to 3016
+    * `:mel_frequency_spacing` - the distance in Hz between two MEL bins before log scaling. Defaults to 66.6
+    * `:type` - Target output type. Defaults to `{:f, 32}`
+
+  ## Examples
+
+      iex> NxSignal.mel_filters(10, 5, 8.0e3)
+      #Nx.Tensor<
+        f32[mels: 5][frequencies: 10]
+        [
+          [0.0, 8.129207999445498e-4, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 9.972016559913754e-4, 2.1870288765057921e-4, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 0.0, 9.510891977697611e-4, 4.150509194005281e-4, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 0.0, 0.0, 4.035891906823963e-4, 5.276656011119485e-4, 2.574124082457274e-4, 0.0, 0.0, 0.0, 0.0],
+          [0.0, 0.0, 0.0, 0.0, 7.329034269787371e-5, 2.342205698369071e-4, 3.8295105332508683e-4, 2.8712040511891246e-4, 1.9128978601656854e-4, 9.545915963826701e-5]
+        ]
+      >
+  """
+  deftransform mel_filters(nfft, nmels, fs, opts \\ []) do
+    opts =
+      Keyword.validate!(opts,
+        max_mel: 3016,
+        mel_frequency_spacing: 200 / 3,
+        type: {:f, 32}
+      )
+
+    mel_filters_n(fs, opts[:max_mel], opts[:mel_frequency_spacing],
+      type: opts[:type],
+      nfft: nfft,
+      nmels: nmels
+    )
+  end
+
+  defnp mel_filters_n(fs, max_mel, f_sp, opts \\ []) do
+    nfft = opts[:nfft]
+    nmels = opts[:nmels]
+    type = opts[:type]
+
+    fftfreqs = fft_frequencies(fs, type: type, nfft: nfft)
+
+    mels = Nx.linspace(0, max_mel / f_sp, type: type, n: nmels + 2, name: :mels)
+    freqs = f_sp * mels
+
+    min_log_hz = 1_000
+    min_log_mel = min_log_hz / f_sp
+
+    # numpy uses the f64 value by default
+    logstep = Nx.log(6.4) / 27
+
+    log_t = mels >= min_log_mel
+
+    # This is the same as freqs[log_t] = min_log_hz * Nx.exp(logstep * (mels[log_t] - min_log_mel))
+    # notice that since freqs and mels are indexed by the same conditional tensor, we don't
+    # need to slice either of them
+    mel_f = Nx.select(log_t, min_log_hz * Nx.exp(logstep * (mels - min_log_mel)), freqs)
+
+    fdiff = Nx.new_axis(mel_f[1..-1//1] - mel_f[0..-2//1], 1)
+    ramps = Nx.new_axis(mel_f, 1) - fftfreqs
+
+    lower = -ramps[0..(nmels - 1)] / fdiff[0..(nmels - 1)]
+    upper = ramps[2..(nmels + 1)//1] / fdiff[1..nmels]
+    weights = Nx.max(0, Nx.min(lower, upper))
+
+    enorm = 2.0 / (mel_f[2..(nmels + 1)] - mel_f[0..(nmels - 1)])
+
+    weights * Nx.new_axis(enorm, 1)
+  end
+
+  @doc """
+  Converts a given STFT time-frequency spectrum into a MEL-scale time-frequency spectrum.
+
+  See also: `stft/3`, `istft/3`, `mel_filters/1`
+
+  ## Arguments
+    * `z` - STFT spectrum
+    * `fs` - Sampling frequency in Hz
+
+  ## Options
+
+    * `:nfft` - Number of FFT bins
+    * `:nmels` - Number of target MEL bins. Defaults to 128
+    * `:type` - Target output type. Defaults to `{:f, 32}`
+
```

Comment:
What about more explicit names, like:

* `fs` -> `sampling_frequency`/`sampling_rate`
* `nfft` -> `num_fft_bins`
* `nmels` -> `num_mel_bins`

Reply:
fs and nfft specifically are usual names for these params, that's why I chose them, but nmels not so much

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html
https://pytorch.org/docs/stable/generated/torch.stft.html
https://www.mathworks.com/help/signal/ref/stft.html

WDYT about:
nfft  -> fft_length
fs -> sampling_frequency
nmels -> mel_bins

Reply:
nfft/fs would need to change for more functions I think

Reply:
FFT is itself an acronym, so adding n on top of that may easily confuse someone without much prior experience, so a more explicit name is generally helpful.

The names above sound good to me!

Reply:
Cool, I'll change these!

Reply:
for consistency, I'm also changing overlap_size -> overlap_length given that both it and nfft operate on the same headspace of "samples"